### ISSUE TITLE
[Bug]: Realtime STT (Voxtral realtime): self-sustained blank-token rut mutes live transcription for minutes over real speech

### ISSUE THREAD

**Issue by damienlaine** [vllm-project/vllm#47614]
## Summary

A realtime transcription session (`/v1/realtime`, Voxtral Mini 4B Realtime, realtime STT path from #45833) can fall into a self-sustained decoding rut: the model emits the blank token (id 32) at every frame (12.5 tok/s) over real speech, for 1 to 20 minutes, then recovers on its own. The engine is perfectly healthy during the episode: steady 1 token/frame, no warnings, KV stable, WebSocket connected. The client just receives empty deltas while speech is flowing, and the audio of the hole is lost.

We reproduce it deterministically on a fixed audio file and have characterized the mechanism. A mitigation PR is linked below.

## Environment

- Realtime STT path as proposed in #45833 (observed on the linto-ai fork it comes from, at the commit matching the PR)
- Model: mistralai/Voxtral-Mini-4B-Realtime-2602, --tokenizer-mode mistral
- Observed on 3 GPUs (RTX A4000, L40S, RTX 4090 laptop), 3 different audio contents (2 audiobooks, 1 live radio stream), mono and multi-session, greedy and temperature 0.3, with and without long-session re-anchoring.

## Deterministic reproduction

Two concurrent realtime WebSocket sessions streaming the same public-domain French audiobook (LibriVox, "Le tour du monde en quatre-vingts jours", resampled 16 kHz mono) at 1x from offsets 2700 s and 5400 s, 30 ms paced appends.

Server:

```
vllm serve mistralai/Voxtral-Mini-4B-Realtime-2602 --tokenizer-mode mistral \
  --max-num-batched-tokens 256 --max-num-seqs 12 --no-enable-prefix-caching \
  --hf-overrides '{"text_config":{"sliding_window":256},"audio_config":{"sliding_window":256}}' \
  --gpu-memory-utilization 0.82 --max-model-len 8192
```

The offset-5400 session goes mute at feed time 1183 s (mid-sentence), stays mute for 179 s (2236 consecutive blank frames), and recovers by itself. Reproduced 8/8, onset stable to 0.1 s, recovery token identical across runs. Trap positions are content-specific: a different long recording will have its own, at other positions.

Harness (self-contained, websockets + soundfile + numpy):

<details>
<summary>collapse_probe.py</summary>

```python
#!/usr/bin/env python3
"""Silent-collapse repro bench for the local 4090 (direct WS, no transcriber).

Streams N desynced channels of the audiobook into the realtime server, with
staggered admissions (the most frequent trigger seen on the L40S run). Each
channel logs non-empty deltas and raises a GAP alarm when no non-empty delta
arrived for GAP_S seconds while audio keeps being fed. Empty deltas received
during the gap are counted: empties flowing = session alive = rut signature
(as opposed to a dead/disconnected session).

Usage: collapse_probe.py [initial_channels] [max_channels] [admit_every_s]
Stop with Ctrl-C / SIGTERM: prints a final gap report.
"""
import asyncio, base64, json, os, signal, sys, time
import numpy as np, soundfile as sf, websockets

URI = os.environ.get("PROBE_URI", "ws://localhost:8000/v1/realtime")
MODEL = "mistralai/Voxtral-Mini-4B-Realtime-2602"
AUDIO = os.environ.get("PROBE_AUDIO", "audiobook_16k_mono.wav")
SR = 16000
FRAME_MS = 30
GAP_S = 60          # no non-empty delta for this long => gap
STATUS_EVERY_S = 60
CHANNEL_OFFSET_S = float(os.environ.get("PROBE_CHANNEL_OFFSET_S", "2700"))  # spacing between channel start positions
# Added to every channel's start position (solo replay of a known rut trigger)
BASE_OFFSET_S = float(os.environ.get("PROBE_BASE_OFFSET_S", "0"))
# If set, append every non-empty delta to <dir>/chNN_transcript.txt (for WER)
DUMP_DIR = os.environ.get("PROBE_DUMP_DIR", "")

N_INIT = int(sys.argv[1]) if len(sys.argv) > 1 else 4
N_MAX = int(sys.argv[2]) if len(sys.argv) > 2 else 10
ADMIT_EVERY_S = float(sys.argv[3]) if len(sys.argv) > 3 else 600.0

T0 = time.time()
def log(msg):
    print(f"[{time.strftime('%H:%M:%S')}][+{time.time()-T0:8.1f}s] {msg}", flush=True)

GAPS = []  # closed gaps: dicts(ch, onset_wall, dur_s, empties, fed_s_at_onset)

class Channel:
    def __init__(self, idx):
        self.idx = idx
        self.tag = f"ch{idx:02d}"
        self.fed_s = 0.0
        self.words = 0
        self.deltas_any = 0
        self.deltas_empty = 0
        self.last_nonempty = None   # wall time of last non-empty delta
        self.last_any = None
        self.gap_open = None        # dict while a gap is open
        self.connected = False
        self.tail = ""
        self.dump_fh = open(f"{DUMP_DIR}/{self.tag}_transcript.txt", "a",
                            buffering=1) if DUMP_DIR else None

    async def run(self):
        f = sf.SoundFile(AUDIO)
        f.seek(int((BASE_OFFSET_S + self.idx * CHANNEL_OFFSET_S) * SR) % max(1, len(f) - SR))
        frame = int(FRAME_MS / 1000 * SR)
        try:
            async with websockets.connect(URI, max_size=None, ping_interval=20) as ws:
                hello = json.loads(await ws.recv())
                log(f"{self.tag} ADMITTED id={hello.get('id')} book_offset={BASE_OFFSET_S + self.idx*CHANNEL_OFFSET_S:.0f}s")
                await ws.send(json.dumps({"type": "session.update", "model": MODEL}))
                await ws.send(json.dumps({"type": "input_audio_buffer.commit"}))
                self.connected = True
                start = time.time()
                recv_task = asyncio.create_task(self.receiver(ws))
                next_t = time.time()
                while True:
                    data = f.read(frame, dtype="float32")
                    if data.ndim > 1: data = data[:, 0]
                    if len(data) == 0:
                        f.seek(0); continue
                    pcm = (data * 32767).astype(np.int16).tobytes()
                    await ws.send(json.dumps({"type": "input_audio_buffer.append",
                                              "audio": base64.b64encode(pcm).decode()}))
                    self.fed_s += len(data) / SR
                    self.check_gap()
                    next_t += FRAME_MS / 1000
                    slack = next_t - time.time()
                    if slack > 0: await asyncio.sleep(slack)
        except Exception as e:
            log(f"{self.tag} DIED: {type(e).__name__}: {e}")
        finally:
            self.connected = False

    async def receiver(self, ws):
        while True:
            try:
                msg = json.loads(await ws.recv())
            except websockets.ConnectionClosed:
                log(f"{self.tag} WS CLOSED by server")
                return
            t = msg.get("type")
            if t == "transcription.delta":
                self.deltas_any += 1
                self.last_any = time.time()
                d = msg.get("delta", "")
                if d.strip():
                    self.words += len(d.split())
                    if self.dump_fh:
                        self.dump_fh.write(d)
                    self.tail = (self.tail + d)[-70:]
                    if self.gap_open:
                        self.close_gap(d)
                    self.last_nonempty = time.time()
                else:
                    self.deltas_empty += 1
                    if self.gap_open:
                        self.gap_open["empties"] += 1
            elif t == "error":
                log(f"{self.tag} SERVER ERROR: {msg.get('error')}")

    def check_gap(self):
        if self.last_nonempty is None or self.gap_open:
            return
        age = time.time() - self.last_nonempty
        if age > GAP_S:
            self.gap_open = {"ch": self.tag, "onset_wall": self.last_nonempty,
                             "empties": 0, "fed_s_at_onset": self.fed_s}
            log(f"{self.tag} *** GAP ONSET *** no non-empty delta for {age:.0f}s "
                f"(fed={self.fed_s:.0f}s, last text: ...{self.tail!r})")

    def close_gap(self, recovery_delta):
        g = self.gap_open
        dur = time.time() - g["onset_wall"]
        g["dur_s"] = round(dur, 1)
        GAPS.append(g)
        log(f"{self.tag} *** GAP END *** duration={dur:.0f}s empties_during={g['empties']} "
            f"recovered_with={recovery_delta!r}")
        self.gap_open = None

async def status_loop(channels):
    while True:
        await asyncio.sleep(STATUS_EVERY_S)
        now = time.time()
        parts = []
        for c in channels:
            if not c.connected:
                parts.append(f"{c.tag}:off")
                continue
            ne_age = f"{now-c.last_nonempty:.0f}s" if c.last_nonempty else "never"
            mark = " GAP!" if c.gap_open else ""
            parts.append(f"{c.tag}:fed={c.fed_s:.0f}s w={c.words} last_txt={ne_age}{mark}")
        log(f"STATUS {' | '.join(parts)}")
        log(f"STATUS gaps_closed={len(GAPS)} open={sum(1 for c in channels if c.gap_open)}")

async def main():
    channels = []
    tasks = []

    def report(sig=None, frame=None):
        log("=" * 70)
        log(f"FINAL REPORT: {len(GAPS)} closed gaps, "
            f"{sum(1 for c in channels if c.gap_open)} still open")
        for g in GAPS:
            log(f"  {g['ch']} dur={g['dur_s']}s empties={g['empties']} fed_at_onset={g['fed_s_at_onset']:.0f}s")
        for c in channels:
            if c.gap_open:
                log(f"  {c.tag} OPEN since fed={c.gap_open['fed_s_at_onset']:.0f}s empties={c.gap_open['empties']}")
        sys.exit(0)

    signal.signal(signal.SIGTERM, report)
    signal.signal(signal.SIGINT, report)

    for i in range(N_INIT):
        c = Channel(i); channels.append(c)
        tasks.append(asyncio.create_task(c.run()))
        await asyncio.sleep(2)
    tasks.append(asyncio.create_task(status_loop(channels)))
    log(f"ramp: {N_INIT} initial channels, +1 every {ADMIT_EVERY_S:.0f}s up to {N_MAX}")

    n = N_INIT
    while n < N_MAX:
        await asyncio.sleep(ADMIT_EVERY_S)
        c = Channel(n); channels.append(c)
        tasks.append(asyncio.create_task(c.run()))
        n += 1

    await asyncio.gather(*tasks)

asyncio.run(main())
```

</details>

Run: `PROBE_URI=ws://localhost:8000/v1/realtime PROBE_BASE_OFFSET_S=2700 python collapse_probe.py 2 2 9999`. The GAP alarm fires when a channel gets no non-empty delta for 60 s while audio keeps being fed; empty deltas are counted during the gap (12.5/s = rut signature, as opposed to a dead session).

## What we measured

1. The rut is token id 32 exclusively, emitted 1/frame. Healthy silences produce blank runs of 25-165 frames; the pathology is run length (2200+).
2. Not sampling-related: identical episode at temperature 0.3 (sampling verified active). The distribution genuinely collapses, see margins below.
3. Not load-related: reproduces at 25 tok/s total on an otherwise idle GPU.
4. Not scheduler-budget-related, but execution details select whether a session is captured: a single session with unsplit encoder chunks never ruts on our reproducer; a single session with split chunks (budget 128 < chunk size ~390 tokens) ruts; two co-scheduled sessions rut at any budget (128, 256, 512, 1024). Any deviation from the solo-unsplit execution path can tip a marginal segment in.
5. Logit margins (per-step logit(32) minus max(other), traced in the model runner): during the rut the blank token wins by +3.5/+5.2/+6.6 logits (p10/p50/p90); on genuinely silent audio it wins by +11.5/+13.9, minimum +8.5; during healthy speech it loses by 11.6 (median). The trapped segments are marginal audio: blank sometimes comes within 0.4 logits of winning even while text flows.
6. Mechanism, consistent with a greedy fixed point: the realtime path decodes at temperature 0.0 and re-feeds generated tokens into the context with a 256-token sliding window. Once the visible context fills with blanks over marginal audio, the state self-sustains until a strong speech onset. Same family as Whisper repetition loops, on the blank token.

## Mitigation

Proposed in #47615: opt-in progressive penalty on the blank token after K consecutive emissions, default off, calibrated from the margins above (the gap between rut margins and true-silence margins makes a safe operating window). Validated on the reproducer: episode cut from 179 s to 18 s, no effect on healthy channels, zero text invented on pure silence, and a 2.5 h 5-stream regression with no hole > 60 s (baseline: 36 holes) at word-level similarity 0.91-0.94 against ground truth.

One architectural finding from the implementation, relevant beyond this issue: the realtime streaming path recycles the engine request for every audio chunk and clears its output-token list in place, so v1 logits processors that rely on `output_tok_ids` from BatchUpdate (min_tokens included) are blind on this path. The penalty is therefore accumulated in the model runner, keyed by request id.

We will file the co-scheduled `/v1/chat/completions` EngineCore crash (inputs_embeds size mismatch in `_preprocess`, kills every realtime session) as a separate issue.
